Makefile

```diff
@@ -240,11 +240,13 @@ generate-go-core: $(CONTROLLER_GEN) $(CONVERSION_GEN)
 		paths=./cmd/clusterctl/...
 	$(MAKE) clean-generated-conversions SRC_DIRS="./api/v1alpha3,./$(EXP_DIR)/api/v1alpha3,./$(EXP_DIR)/addons/api/v1alpha3"
 	$(CONVERSION_GEN) \
+		--output-base=./ \
```

Comment:
Might be I'm getting this wrong, but this means we are putting files in the project root? Shot the be ./api/v1alpha3 instead?
Also, what about moving this flag near --output-file-base, so we have all the flags determining our output  one near the others?

Reply:
This should only ask the conversion tool to base the output in the current folder, relative to the Makefile (in this instance), which makes sense. 

That said, when running it locally I get this panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x12bbe0d]

goroutine 1 [running]:
k8s.io/gengo/parser.(*Builder).AddDirTo(0xc0000b5680, 0xc00079a414, 0x24, 0xc0003e2008, 0x0, 0x0)
	/Users/vince/go/pkg/mod/k8s.io/gengo@v0.0.0-20201214224949-b6c5ce23f027/parser/parse.go:272 +0xad
k8s.io/gengo/generator.(*Context).AddDir(...)
	/Users/vince/go/pkg/mod/k8s.io/gengo@v0.0.0-20201214224949-b6c5ce23f027/generator/generator.go:246
k8s.io/code-generator/cmd/conversion-gen/generators.Packages(0xc0003e2000, 0xc0000d4fa0, 0x13753bb, 0x6, 0xc0003e2000)
	/Users/vince/go/pkg/mod/k8s.io/code-generator@v0.21.0-beta.0/cmd/conversion-gen/generators/conversion.go:307 +0x1289
k8s.io/gengo/args.(*GeneratorArgs).Execute(0xc0000d4fa0, 0xc0000c1080, 0x13753bb, 0x6, 0x13a3608, 0xc00007e660, 0x1)
	/Users/vince/go/pkg/mod/k8s.io/gengo@v0.0.0-20201214224949-b6c5ce23f027/args/args.go:206 +0x1b7
main.main()
	/Users/vince/go/pkg/mod/k8s.io/code-generator@v0.21.0-beta.0/cmd/conversion-gen/main.go:129 +0x365
```

Reply:
It does make it look like that, but the output-base is actually just the "base" where things go, then the files themselves are placed relative to that base. So when GOPATH is set, that causes the path to be $GOPATH/src/api/v1alpha3, etc. With this set to ./, the file path is where we want it under $REPO_ROOT/api/v1alpha3.

I can change the grouping if you prefer.

Reply:
Hmm, that's very odd @vincepri. Any interesting environment information that could help explain this?

One thing I ran into testing elsewhere was having an old `conversion-gen` binary that had been previously created. During the build, we check for the file's presence and build it if needed. But once that is done initially, it doesn't ever touch it again. It might be interesting to try `rm -fr hack/tools/bin/*` and run again to see if you still get that error.

Reply:
This did not work for me as well...
I have `GOPATH=/Users/fpandini/go`, I run make generate from `/Users/fpandini/go/src/sigs.k8s.io/cluster-api` and generated file got removed

Reply:
I'm not sure I understand. Is your local clone under `/Users/fpandini/go/src` or elsewhere?

I have not tried actually running with the clone under GOPATH, but I will try that now. From the argument description, that should not really make a difference, but it's possible there's something else wrong there.

I haven't been able to reproduce the CI failures either.

Reply:
@stmcginnis What is the version of go that you have installed?
I am seeing the same error that @vincepri is seeing using `go version go1.16.2 darwin/amd64`

Reply:
FWIW:
1. I have $GOPATH set to `/Users/ashisham/go`
2. Code coned at `/Users/ashisham/work/kubernetes/src/code/cluster-api`

Reply:
I'm running go1.16.2, but I've also tested with go1.15.x. That was my initial assumption. Then I saw a difference between Linux and macOS. But finally I noticed the differences in environments appeared to be whether GOPATH was set or not.

That appears to have been the issue in some cases, but the actual code location appears to be another variable in the equation. I was able to observe differences whether the code was cloned into any random filesystem path versus when it was actually cloned under GOPATH.

The link Cecile pointed to in CAPZ looks like it may be the key, with only setting if the code is actually not under GOPATH. I've run some local tests, and it looks like that should do the trick. That code has been in CAPZ for several months now, so presumable that does the trick.

Just a couple more scenarios to try locally here, then I will have an update to push up. Thanks for the additional context, that does help!